File: ios/iOSBaseProject/Domain/UseCases/Auth/SignInUseCase.swift
//
//  SignInUseCase.swift
//  BaatCheet
//
//  Domain layer - Sign In Use Case
//

import Foundation

/// Use case for signing in a user.
/// Validates credentials and delegates to the repository.
protocol SignInUseCaseProtocol {
    func execute(email: String, password: String) async -> AuthResult
}

final class SignInUseCase: SignInUseCaseProtocol {
    private let repository: AuthRepositoryProtocol
    
    init(repository: AuthRepositoryProtocol) {
        self.repository = repository
    }
    
    func execute(email: String, password: String) async -> AuthResult {
        // Validate inputs
        guard !email.isEmpty else {
            return .failure(.invalidEmail)
        }
        
        guard !password.isEmpty else {
            return .failure(.weakPassword)
        }
        
        guard password.count >= 8 else {
            return .failure(.weakPassword)
        }
        
        // Validate email format
        let credentials = AuthCredentials(email: email, password: password)
        guard credentials.isEmailValid else {
            return .failure(.invalidEmail)
        }
        
        // Delegate to repository
        return await repository.signIn(email: email, password: password)
    }
}

File: android/app/src/main/java/com/baatcheet/app/domain/usecase/auth/SignInUseCase.kt
package com.baatcheet.app.domain.usecase.auth

import com.baatcheet.app.domain.model.AuthResult
import com.baatcheet.app.domain.repository.AuthRepository
import javax.inject.Inject

/**
 * Sign In Use Case
 * 
 * Handles the business logic for user sign-in.
 * This use case validates credentials and delegates to the repository.
 */
class SignInUseCase @Inject constructor(
    private val authRepository: AuthRepository
) {
    /**
     * Execute sign in with email and password
     */
    suspend operator fun invoke(email: String, password: String): AuthResult {
        // Validate inputs
        if (email.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Email cannot be empty"))
        }
        if (password.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Password cannot be empty"))
        }
        if (password.length < 8) {
            return AuthResult.Failure(IllegalArgumentException("Password must be at least 8 characters"))
        }
        
        return authRepository.signIn(email, password)
    }
}
